### DISCUSSION TITLE
How to handle management columns that cozemble requires in every table

### DISCUSSION THREAD

**Discussion by mike-hogan** [cozemble/monorepo#34] [Ideas]
Let's say you model a Customer in cozemble, with First Name, Last Name, Email and Phone Number.  Then tell it to create the associated database schema.

You would expect to only see a `customer` table with `first_name, last_name, email and phone_number` columns.

But cozemble needs more to be able to do its thing.  It will need

 * record_id
 * created_by
 * created_at
 * updated_at
 * is_deleted (to support soft deletion)

And possibly others as things evolve

Adding these as columns to the table will pollute any auto-generated graphql or rest api.  Altho this is probably work-aroundable, via configuration or what have you

There is also the smaller issue of possible name clashes with column names the user wants in their model.  This can be worked around by prefixing all of the above fields with _coz_.

One other alternative is just to bundle all these `coz` fields into a single `jsonb` called `_coz_metadata`.  This has its own drawbacks, or jsonb support in gql apis for example.  Supabase right now does not seem to support that, but it will.  Also its is slightly harder to update and query these nested jsonb fields.

Super unclear on what approach is best.

So the two options are:

1. fields in the table called `_coz_record_id, _coz_created_by, _coz_is_deleted` etc
2. one metadata field called `_coz_metadata`

After writing that I just thought of another.  Every table could have an associated has-one metadata table.

So if I have a Customer model, I will have a `customer` table and a `customer_coz_metadata` table, and the `customer` table will have a mandatory foreign key to its metadata table, and a graphql relationship created also.  Will explore this option.

3. have a has-one metadata table for every model generated table

**Comment by Metehan-Altuntekin** [comment]
The 3rd option doesn't seem convenient to me. UI will need some of those fields, for example: `created_at` for sorting. Since there will be a need of adding a property to the record for the key anyway, it may be more of a gimmick than to be a utility.

The 1st and 2nd options seem good, keeping them in a single object is even better with the 2nd. 

I am not much familiar with Postgres or SQL databases in general, so I don't know what is the reason jsonb is preferred than ordinary nesting.
